### PR TITLE
Add a lesson about regular expressions

### PR DESCRIPTION
**Pull request by sebastiangeiger**
- [ ] Address @iamvery's remarks
- [ ] Add tests once #21 is merged

### CODE REVIEW

**Comment by wolfbrain** [comment]
:eyes: @iamvery

File: about_regular_expressions.exs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+defmodule AboutRegularExpressions do
+  use Koans
+
+  think "A simple regular expression" do
+    regex = ~r/This [W|w]orks!/
+
+    completion = __?
+
+    assert ("This W" <> completion) =~ regex
```
Comment (iamvery) [inline about_regular_expressions.exs:9]: I'm not sure I like the "completion" being the missing piece. I'd say maybe just leave out the whole string. What about

``` elixir
think "A simple regular expression" do
  regex = ~r/This [W|w]orks!/

  assert __? =~ regex, "it matches the uppercase W"
  assert __? =~ regex, "it matches the lowercase w"
end
```

File: about_regular_expressions.exs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    assert ("This w" <> completion) =~ regex
+  end
+
+  think "Regular expressions can use options" do
```
Comment (iamvery) [inline about_regular_expressions.exs:13]: I believe the correct terminology is s/options/modifiers/

File: about_regular_expressions.exs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    regex = ~r/this works!/i
```
Comment (iamvery) [inline about_regular_expressions.exs:14]: perhaps name this `case_insensitive_regex`?

File: about_regular_expressions.exs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    assert_? "this works!" =~ regex
+    assert_? "ThIs wOrKs!" =~ regex
+  end
+
+  think "Regular expressions don't have to be created with the sigil (though ~r/.../ is more idiomatic)" do
```
Comment (iamvery) [inline about_regular_expressions.exs:20]: I might reword this to say something like

> Regular expressions may be dynamically compiled from a string

File: about_regular_expressions.exs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    regex = Regex.compile!(__?)
+
+    assert regex == ~r/this works!/
```
Comment (iamvery) [inline about_regular_expressions.exs:23]: Should this be `=~`?

Comment (iamvery) [inline about_regular_expressions.exs:23]: oh, right, it's comparing the expressions as values... maybe actually use the regex instead?

File: about_regular_expressions.exs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  think "Sigils have flexible delimiters" do
```
Comment (iamvery) [inline about_regular_expressions.exs:26]: I might reword as

> Remember: sigils may use different delimiters

File: about_regular_expressions.exs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    assert __? == ~r(this works!)
+  end
+
+  think "Capture groups" do
+    regex = ~r/(A|The) lazy bear slept for (\d+) months/
+    text = "The lazy bear slept for 5 months"
+    assert Regex.run(regex, text) == __?
```
Comment (iamvery) [inline about_regular_expressions.exs:33]: We might want to give them a hint about the structure that is returned.

``` elixir
assert Regex.run(regex, text) == [__?, __?]
```

Also I noticed that there are some additional options to this function (e.g. `return: :index`). Should we illustrate that usage as well?

File: about_regular_expressions.exs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  think "Replacing with capture groups" do
+    text = "This has vowels"
+    regex = ~r/([aeiou])/
+    assert Regex.replace(regex, text, "<\\1>") == __?
```
Comment (iamvery) [inline about_regular_expressions.exs:39]: Based on the docs there are lot's of flavors to this function.
- option `global: false`
- 3rd arg function to handle replacement

Also what is the significance of the format of the replacement? I assume `\\1` is referencing the first capture group and is `\\` to escape the literal `\`. What are the angle brackets for? Can/should we want to use a sigil to avoid having to escape the backslash?

**Comment by iamvery** [comment]
@sebastiangeiger were you planning to respond here, or do you want me to take over?

**Comment by sebastiangeiger** [comment]
@iamvery I'll try and finish it before I go on vacation next week

**Comment by sebastiangeiger** [comment]
@iamvery I didn't have time to work on it. Feel free to take it over, otherwise I'll get to it once I'm back from vacation and all caught up again.

**Comment by iamvery** [comment]
No problem! You may have caught on that I'm favoring this new project a bit elixirkoans.io
